File: Data Structures & Algorithms/max-water-container/submission-5.kt
import kotlin.math.max
import kotlin.math.min

class Solution {
    fun maxArea(heights: IntArray): Int {
        var left = 0
        var right = heights.lastIndex

        var maxArea = 0

        while (left < right) {
            val currentMinHeight = min(heights[left], heights[right])
            val base = right - left
            val currentArea = currentMinHeight * base
            maxArea = max(maxArea, currentArea)

            when {
                heights[left] > heights[right] -> right--
                heights[left] < heights[right] -> left++
                else -> left++
            }
        }

        return maxArea
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-5.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        left = 0
        right = 1

        while left < right and right < len(prices):
            if prices[left] < prices[right]:
                buy = prices[left]
                sell = prices[right]
                diff = sell - buy
                max_profit = max(max_profit, diff)
            else:
                left = right

            right += 1

        return max_profit